### DISCUSSION TITLE
TrustyAI integration with Guardrails

### DISCUSSION THREAD

**Discussion by evaline-ju** [trustyai-explainability/community#40] [Ideas]
“Guardrails” is a stack of various components that enable standalone detections on text, as well as invoking detections on text generation server input (i.e. user prompts) and text generation server output (i.e. generated text).

## Background

Components of the "Guardrails" stack include:
- Orchestrator - available [open-sourced here](https://github.com/foundation-model-stack/fms-guardrails-orchestrator) / [API here](https://foundation-model-stack.github.io/fms-guardrails-orchestrator/)
     - This REST service is the main entry point for a user trying to get detection information, such as on their LLM inputs and/or outputs. It coordinates the calls to text generation servers (LLM servers), chunker server(s), and detector server(s), described below.
- Detectors / detector servers
     - These REST services provide detection information such as (but not limited to) classifications on text and are expected to conform to the [detectors API](https://foundation-model-stack.github.io/fms-guardrails-orchestrator/?urls.primaryName=Detector+API). 
- Chunkers / chunker servers
     - Chunkers can be thought of as servers that “chunk” the target modality. For text, this can look like tokenizers or sentence splitters. Currently to work with the orchestrator, these are expected to be GRPC servers, to support bidirectional streaming. Detectors can specify corresponding chunkers in configuration. 
     - [Proto for reference](https://github.com/foundation-model-stack/fms-guardrails-orchestrator/blob/main/protos/caikit_runtime_Chunkers.proto)

We generally have assumed the text generation server(s) are separate from the “Guardrails” stack i.e. deployed and maintained separately. The grpc text generation provider can be specified through orchestrator configuration. Currently, the orchestrator is directly compatible and tested with:
1.  TGIS grpc interface, [proto](https://github.com/IBM/text-generation-inference/blob/main/proto/generation.proto)
2.  [caikit-nlp](https://github.com/caikit/caikit-nlp) text generation interface

## Integration

This integration will allow for TrustyAI users to be able to get detections in their LLM applications in real-time and allow users to bring in other detectors (e.g. potentially toxic text detectors) into their applications.

NOTE: This particular TrustyAI integration discussion will be considered separately of being able to build and provide official images for the open-sourced [orchestrator](https://github.com/foundation-model-stack/fms-guardrails-orchestrator), along with any image maintenance issues such as security scanning, and thus is not contingent on Rust knowledge.

The integration will include:
- Standing up an open-sourced Guardrails stack
     - This will require open-sourced versions of at least one running detector and potentially chunker, depending on the detector of choice.
- Controller to be able to deploy / maintain a Guardrails stack via the TrustyAI operator
    - This will eventually allow RHOAI consumers access to the Guardrails stack. However, we want to make sure the features are toggle-able so a consumer who may want only metrics [through TrustyAIService today] or only detections are not bound to include other features.

Starting this discussion to make sure the points of intended integration are agreed upon, and if so, potentially the details of moving forward

**Comment by ruivieira** [comment]
@evaline-ju thank you for the detailed proposal, this looks great!

Can I suggest that this is also submitted as an ADR (could be verbatim) to the [community repo](https://github.com/trustyai-explainability/community/tree/main/adr)? This way we can formalise the planning around this effort.

**Reply by evaline-ju** [reply]
sure, I can contribute an ADR

**Reply by evaline-ju** [reply]
ADR PR added: https://github.com/trustyai-explainability/community/pull/41

**Comment by ruivieira** [comment]
@evaline-ju with the inclusion of the orchestrator in the upcoming TrustyAI operator 1.33 release, I suggest closing this discussion and keep it for reference. We can open more specific discussion threads if needed. Do you agree?

**Reply by evaline-ju** [reply]
@ruivieira Yes, keeping as reference works for me - thanks!
